File: android-client/app/src/main/java/io/s3s5/android/crypto/Codec.kt
package io.s3s5.android.crypto

import org.json.JSONObject
import java.security.GeneralSecurityException
import java.security.SecureRandom
import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

interface Codec {
    fun seal(objectType: String, sessionId: String, direction: String, seq: Long, plaintext: ByteArray): ByteArray
    fun open(objectType: String, sessionId: String, direction: String, seq: Long, data: ByteArray): ByteArray
    fun sealData(sessionId: String, direction: String, seq: Long, plaintext: ByteArray): ByteArray
    fun openData(sessionId: String, direction: String, seq: Long, data: ByteArray): ByteArray
    fun enabled(): Boolean
}

object NoopCodec : Codec {
    override fun seal(objectType: String, sessionId: String, direction: String, seq: Long, plaintext: ByteArray): ByteArray =
        plaintext.copyOf()

    override fun open(objectType: String, sessionId: String, direction: String, seq: Long, data: ByteArray): ByteArray =
        data.copyOf()

    override fun sealData(sessionId: String, direction: String, seq: Long, plaintext: ByteArray): ByteArray =
        plaintext.copyOf()

    override fun openData(sessionId: String, direction: String, seq: Long, data: ByteArray): ByteArray =
        data.copyOf()

    override fun enabled(): Boolean = false
}

class PskCodec(psk: String) : Codec {
    private val pskBytes = psk.toByteArray(Charsets.UTF_8)
    private val random = SecureRandom()
    private val b64 = Base64.getEncoder()
    private val b64d = Base64.getDecoder()

    init {
        require(psk.length >= 16) { "S3S5_PSK must be at least 16 characters" }
    }

    override fun enabled(): Boolean = true

    override fun seal(
        objectType: String,
        sessionId: String,
        direction: String,
        seq: Long,
        plaintext: ByteArray,
    ): ByteArray {
        val nonce = randomNonce()
        val ciphertext = sealRaw(objectType, sessionId, direction, seq, nonce, plaintext)
        return JSONObject()
            .put("v", 1)
            .put("alg", "AES-256-GCM")
            .put("nonce", b64.encodeToString(nonce))
            .put("ciphertext", b64.encodeToString(ciphertext))
            .toString()
            .toByteArray(Charsets.UTF_8)
    }

    override fun open(
        objectType: String,
        sessionId: String,
        direction: String,
        seq: Long,
        data: ByteArray,
    ): ByteArray {
        val env = JSONObject(String(data, Charsets.UTF_8))
        if (env.getInt("v") != 1 || env.getString("alg") != "AES-256-GCM") {
            throw GeneralSecurityException("unsupported crypto envelope")
        }
        val nonce = b64d.decode(env.getString("nonce"))
        if (nonce.size != 12) {
            throw GeneralSecurityException("invalid crypto nonce size")
        }
        val ciphertext = b64d.decode(env.getString("ciphertext"))
        return openRaw(objectType, sessionId, direction, seq, nonce, ciphertext)
    }

    override fun sealData(sessionId: String, direction: String, seq: Long, plaintext: ByteArray): ByteArray {
        val nonce = randomNonce()
        val ciphertext = sealRaw("data", sessionId, direction, seq, nonce, plaintext)
        require(ciphertext.size <= Int.MAX_VALUE - 24)
        val out = ByteArray(12 + nonce.size + ciphertext.size)
        out[0] = 'S'.code.toByte()
        out[1] = '5'.code.toByte()
        out[2] = 'D'.code.toByte()
        out[3] = '1'.code.toByte()
        out[4] = 1
        out[5] = 1
        out[6] = nonce.size.toByte()
        out[7] = 0
        writeU32(out, 8, ciphertext.size)
        nonce.copyInto(out, 12)
        ciphertext.copyInto(out, 12 + nonce.size)
        return out
    }

    override fun openData(sessionId: String, direction: String, seq: Long, data: ByteArray): ByteArray {
        if (data.size < 24) {
            throw GeneralSecurityException("truncated data crypto envelope")
        }
        if (data[0] != 'S'.code.toByte() || data[1] != '5'.code.toByte() || data[2] != 'D'.code.toByte() || data[3] != '1'.code.toByte()) {
            throw GeneralSecurityException("invalid data crypto envelope magic")
        }
        if (data[4].toInt() != 1 || data[5].toInt() != 1) {
            throw GeneralSecurityException("unsupported data crypto envelope")
        }
        val nonceLen = data[6].toInt() and 0xff
        if (nonceLen != 12) {
            throw GeneralSecurityException("invalid data crypto nonce size")
        }
        val ctLen = readU32(data, 8)
        if (ctLen < 16 || data.size != 12 + nonceLen + ctLen) {
            throw GeneralSecurityException("invalid data crypto envelope length")
        }
        val nonce = data.copyOfRange(12, 12 + nonceLen)
        val ciphertext = data.copyOfRange(12 + nonceLen, data.size)
        return openRaw("data", sessionId, direction, seq, nonce, ciphertext)
    }

    private fun randomNonce(): ByteArray {
        val nonce = ByteArray(12)
        random.nextBytes(nonce)
        return nonce
    }

    private fun sealRaw(objectType: String, sessionId: String, direction: String, seq: Long, nonce: ByteArray, plaintext: ByteArray): ByteArray {
        val cipher = newCipher(Cipher.ENCRYPT_MODE, sessionId, direction, nonce)
        cipher.updateAAD(associatedData(objectType, sessionId, direction, seq))
        return cipher.doFinal(plaintext)
    }

    private fun openRaw(objectType: String, sessionId: String, direction: String, seq: Long, nonce: ByteArray, ciphertext: ByteArray): ByteArray {
        val cipher = newCipher(Cipher.DECRYPT_MODE, sessionId, direction, nonce)
        cipher.updateAAD(associatedData(objectType, sessionId, direction, seq))
        return cipher.doFinal(ciphertext)
    }

    private fun newCipher(mode: Int, sessionId: String, direction: String, nonce: ByteArray): Cipher {
        val key = deriveKey(sessionId, direction)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(mode, SecretKeySpec(key, "AES"), GCMParameterSpec(128, nonce))
        return cipher
    }

    private fun deriveKey(sessionId: String, direction: String): ByteArray =
        hkdfSha256(
            secret = pskBytes,
            salt = "s3s5/v1/$sessionId".toByteArray(Charsets.UTF_8),
            info = "payload/$direction".toByteArray(Charsets.UTF_8),
            n = 32,
        )

    companion object {
        fun associatedData(objectType: String, sessionId: String, direction: String, seq: Long): ByteArray =
            "s3s5/v1|$objectType|$sessionId|$direction|${"%020d".format(seq)}".toByteArray(Charsets.UTF_8)

        fun hkdfSha256(secret: ByteArray, salt: ByteArray, info: ByteArray, n: Int): ByteArray {
            val prk = hmacSha256(salt, secret)
            val out = ArrayList<Byte>(n)
            var prev = ByteArray(0)
            var counter = 1
            while (out.size < n) {
                val mac = Mac.getInstance("HmacSHA256")
                mac.init(SecretKeySpec(prk, "HmacSHA256"))
                mac.update(prev)
                mac.update(info)
                mac.update(counter.toByte())
                prev = mac.doFinal()
                for (b in prev) {
                    if (out.size < n) {
                        out.add(b)
                    }
                }
                counter++
            }
            return ByteArray(out.size) { out[it] }
        }

        private fun hmacSha256(key: ByteArray, data: ByteArray): ByteArray {
            val mac = Mac.getInstance("HmacSHA256")
            mac.init(SecretKeySpec(key, "HmacSHA256"))
            return mac.doFinal(data)
        }

        private fun writeU32(out: ByteArray, offset: Int, value: Int) {
            out[offset] = ((value ushr 24) and 0xff).toByte()
            out[offset + 1] = ((value ushr 16) and 0xff).toByte()
            out[offset + 2] = ((value ushr 8) and 0xff).toByte()
            out[offset + 3] = (value and 0xff).toByte()
        }

        private fun readU32(data: ByteArray, offset: Int): Int =
            ((data[offset].toInt() and 0xff) shl 24) or
                ((data[offset + 1].toInt() and 0xff) shl 16) or
                ((data[offset + 2].toInt() and 0xff) shl 8) or
                (data[offset + 3].toInt() and 0xff)
    }
}

File: internal/crypto/codec.go
package crypto

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"hash"
	"io"
)

type Codec interface {
	Seal(objectType, sessionID, direction string, seq uint64, plaintext []byte) ([]byte, error)
	Open(objectType, sessionID, direction string, seq uint64, data []byte) ([]byte, error)
	SealData(sessionID, direction string, seq uint64, plaintext []byte) ([]byte, error)
	OpenData(sessionID, direction string, seq uint64, data []byte) ([]byte, error)
	Enabled() bool
}

type NoopCodec struct{}

func (NoopCodec) Seal(_, _, _ string, _ uint64, plaintext []byte) ([]byte, error) {
	return append([]byte(nil), plaintext...), nil
}

func (NoopCodec) Open(_, _, _ string, _ uint64, data []byte) ([]byte, error) {
	return append([]byte(nil), data...), nil
}

func (NoopCodec) SealData(_, _ string, _ uint64, plaintext []byte) ([]byte, error) {
	return append([]byte(nil), plaintext...), nil
}

func (NoopCodec) OpenData(_, _ string, _ uint64, data []byte) ([]byte, error) {
	return append([]byte(nil), data...), nil
}

func (NoopCodec) Enabled() bool { return false }

type PSKCodec struct {
	psk []byte
}

type envelope struct {
	Version    int    `json:"v"`
	Algorithm  string `json:"alg"`
	Nonce      string `json:"nonce"`
	Ciphertext string `json:"ciphertext"`
}

var dataMagic = [4]byte{'S', '5', 'D', '1'}

const (
	dataEnvelopeVersion = 1
	dataAlgorithmAESGCM = 1
	dataHeaderLen       = 12
)

func NewPSKCodec(psk string) (*PSKCodec, error) {
	if len(psk) < 16 {
		return nil, errors.New("S3S5_PSK must be at least 16 characters")
	}
	return &PSKCodec{psk: []byte(psk)}, nil
}

func (c *PSKCodec) Enabled() bool { return true }

func (c *PSKCodec) Seal(objectType, sessionID, direction string, seq uint64, plaintext []byte) ([]byte, error) {
	return c.sealJSON(objectType, sessionID, direction, seq, plaintext)
}

func (c *PSKCodec) Open(objectType, sessionID, direction string, seq uint64, data []byte) ([]byte, error) {
	return c.openJSON(objectType, sessionID, direction, seq, data)
}

func (c *PSKCodec) SealData(sessionID, direction string, seq uint64, plaintext []byte) ([]byte, error) {
	nonce, ciphertext, err := c.sealRaw("data", sessionID, direction, seq, plaintext)
	if err != nil {
		return nil, err
	}
	out := make([]byte, dataHeaderLen+len(nonce)+len(ciphertext))
	copy(out[:4], dataMagic[:])
	out[4] = dataEnvelopeVersion
	out[5] = dataAlgorithmAESGCM
	out[6] = byte(len(nonce))
	out[7] = 0
	binary.BigEndian.PutUint32(out[8:12], uint32(len(ciphertext)))
	copy(out[dataHeaderLen:], nonce)
	copy(out[dataHeaderLen+len(nonce):], ciphertext)
	return out, nil
}

func (c *PSKCodec) OpenData(sessionID, direction string, seq uint64, data []byte) ([]byte, error) {
	if len(data) < dataHeaderLen {
		return nil, errors.New("truncated data crypto envelope")
	}
	if !bytes.Equal(data[:4], dataMagic[:]) {
		return nil, errors.New("invalid data crypto envelope magic")
	}
	if data[4] != dataEnvelopeVersion {
		return nil, errors.New("unsupported data crypto envelope version")
	}
	if data[5] != dataAlgorithmAESGCM {
		return nil, errors.New("unsupported data crypto envelope algorithm")
	}
	nonceLen := int(data[6])
	if nonceLen != 12 {
		return nil, errors.New("invalid data crypto nonce size")
	}
	ctLen := int(binary.BigEndian.Uint32(data[8:12]))
	if ctLen < 16 {
		return nil, errors.New("invalid data crypto ciphertext size")
	}
	if len(data) != dataHeaderLen+nonceLen+ctLen {
		return nil, errors.New("invalid data crypto envelope length")
	}
	nonce := data[dataHeaderLen : dataHeaderLen+nonceLen]
	ciphertext := data[dataHeaderLen+nonceLen:]
	return c.openRaw("data", sessionID, direction, seq, nonce, ciphertext)
}

func (c *PSKCodec) sealJSON(objectType, sessionID, direction string, seq uint64, plaintext []byte) ([]byte, error) {
	nonce, ct, err := c.sealRaw(objectType, sessionID, direction, seq, plaintext)
	if err != nil {
		return nil, err
	}
	env := envelope{
		Version:    1,
		Algorithm:  "AES-256-GCM",
		Nonce:      base64.StdEncoding.EncodeToString(nonce),
		Ciphertext: base64.StdEncoding.EncodeToString(ct),
	}
	return json.Marshal(env)
}

func (c *PSKCodec) openJSON(objectType, sessionID, direction string, seq uint64, data []byte) ([]byte, error) {
	var env envelope
	if err := json.Unmarshal(data, &env); err != nil {
		return nil, err
	}
	if env.Version != 1 || env.Algorithm != "AES-256-GCM" {
		return nil, errors.New("unsupported crypto envelope")
	}
	nonce, err := base64.StdEncoding.DecodeString(env.Nonce)
	if err != nil {
		return nil, err
	}
	ct, err := base64.StdEncoding.DecodeString(env.Ciphertext)
	if err != nil {
		return nil, err
	}
	return c.openRaw(objectType, sessionID, direction, seq, nonce, ct)
}

func (c *PSKCodec) sealRaw(objectType, sessionID, direction string, seq uint64, plaintext []byte) ([]byte, []byte, error) {
	key := c.deriveKey(sessionID, direction)
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, nil, err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, nil, err
	}
	nonce := make([]byte, gcm.NonceSize())
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		return nil, nil, err
	}
	aad := associatedData(objectType, sessionID, direction, seq)
	ct := gcm.Seal(nil, nonce, plaintext, aad)
	return nonce, ct, nil
}

func (c *PSKCodec) openRaw(objectType, sessionID, direction string, seq uint64, nonce, ct []byte) ([]byte, error) {
	key := c.deriveKey(sessionID, direction)
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}
	if len(nonce) != gcm.NonceSize() {
		return nil, errors.New("invalid crypto nonce size")
	}
	aad := associatedData(objectType, sessionID, direction, seq)
	return gcm.Open(nil, nonce, ct, aad)
}

func (c *PSKCodec) deriveKey(sessionID, direction string) []byte {
	salt := []byte("s3s5/v1/" + sessionID)
	info := []byte("payload/" + direction)
	return hkdfSHA256(c.psk, salt, info, 32)
}

func associatedData(objectType, sessionID, direction string, seq uint64) []byte {
	return []byte(fmt.Sprintf("s3s5/v1|%s|%s|%s|%020d", objectType, sessionID, direction, seq))
}

func hkdfSHA256(secret, salt, info []byte, n int) []byte {
	prk := hmacHash(sha256.New, salt, secret)
	var out bytes.Buffer
	var prev []byte
	counter := byte(1)
	for out.Len() < n {
		h := hmac.New(sha256.New, prk)
		h.Write(prev)
		h.Write(info)
		h.Write([]byte{counter})
		prev = h.Sum(nil)
		out.Write(prev)
		counter++
	}
	return out.Bytes()[:n]
}

func hmacHash(h func() hash.Hash, key, data []byte) []byte {
	m := hmac.New(h, key)
	m.Write(data)
	return m.Sum(nil)
}
